R0 correction

abriz97 opened the pull request:
**Pull Request**

Thanks for your contribution to EpiEstim! Please include the following information:

* A description of this PR, and why it is needed.
I implement the back-imputation procedure described in Brizzi, O'Driscoll and Dorigatti to mitigate early-$R_t$ over-estimation.

* A summary of the changes.
I tried to make as little changes as possible,  adding exclusively one function (`backimpute_I()`, to estimate incidence prior to the first reported case), and one argument to `estimate_R()` (`backimputation_window`, determining the number of observations used for the backimputation step).
By default, `estimate_R` behaves as in the previous versions. 
Currently, `backimpute_I()` does not support objects of class "incidence" as it does not deal with imported cases. 

* Links to any issues it fixes.
Not a github issue, but addresses O'Driscoll et al.'s  [A Comparative Analysis of Statistical Methods to Estimate the Reproduction Number in Emerging Epidemics, With Implications for the Current Coronavirus Disease 2019 Pandemic](https://doi.org/10.1093/cid/ciaa1599)

**How has this been tested**
I locally checked that all past tests passed without errors. I also added tests confirming that warnings and errors work as required and outputs work as expected  

**Checklist**
- [ V ] I have added tests to prove my changes work.
  * tests are available at `tests/testthat/test-backimpute.R`
- [ V ] I have added documentation where required:
  * added documentation for new `backimpute_I()` function and new parameter `backimputation_window` for `estimate_R()` parameter. 
- [ V ] I have updated `NEWS.md` with a short description of my change

sangeetabhatia03 commented on R/backimputation.R:
```diff
@@ -0,0 +1,85 @@
+
+    predict_backimpute_log <- predict.lm(fit_backimpute, newdata = list(t = imputed_t))
+    predict_backimpute <- data.frame(
+        local = exp(predict_backimpute_log) - safe_shift,
```
@abriz97 can you please double check that this is correct and that it should not be exp(safe_shift)? Asking because you did log(incid + safe_shift) in the previous step?

abriz97 replied:
This is indeed the correct way of moving from the "log-incidence" world to the "count-incidence" world.
Mathematically:
safe_log_incid = log(incid + safe_shift)      $\implies$      incid = exp(safe_log_incid) - safe_shift

I find it helpful to think of the "safe_shift" as an additional case per day.

sangeetabhatia03 replied:
Ah of course! i was trying to do it in my head :) Thanks Andrea!

sangeetabhatia03 commented:
@abriz97 @annecori I have reviewed the PR and the implementation looks good to me. I have a small question (left in the comments) but otherwise I have no concerns.

annecori commented:
Hi @abriz97  thanks for this contribution, I'm finally going through this. Can you let me know if you would be happy for me to change the plots in the new vignette so that Rt is plotted at the end of the interval rather than the midpoint? This would ensure consistency with all other plots produced by default as part of the package. Thanks

annecori commented on R/backimputation.R:
```diff
@@ -0,0 +1,86 @@
+    incid_processed[1, ] <- c(sum(incid_processed[1, ]), 0)
+
+    # some cases may be 0, implying -infinite logs
+    safe_shift <- .5
```
@sangeetabhatia03  do we think it's a problem that this is hardcoded? 
@abriz97  is this choice documented well?

abriz97 replied:
I am closing this as I addressed it in the next comment.

annecori commented on R/backimputation.R:
```diff
@@ -0,0 +1,86 @@
+        t = seq(window_b),
+        li = log(incid[1:window_b] + safe_shift)
+   )
+    imputed_t <- seq(from = -100, to = 0)
```
@sangeetabhatia03  and @abriz97  I think it would be best to not barcode -100? Or do we think it's ok?

abriz97 replied:
Hi @annecori, 

This is a good point that deserves discussion. My answer below covers both this and the `safe_shift` question, as your concerns are similar.

In my upcoming pushed commit, I mention the two decisions in the documentation of `backimpute_I()`:

```{R backimpute_I_documentation}
#' Log-incidence is computed as log(local + 0.5) to avoid -infinite logs.
...
#' @return an incidence data.frame, combining back-imputed cases for a maximum 
#' of 100 time points (with rows indexed by a negative integer rowname) and 
#' cases (with rows #' indexed by a non-negative integer)
```

In both cases, there is no mathematical argument justifying the exact choice of `safe_shift=0.5` or the maximum of 100 time units for which the back-imputation is performed. The heuristic justifications are: 
* I chose `safe_shift=0.5` as an "average" between one and zero cases. This would be a "middle ground" between a world where the observed incidence is correct, and a world in which 1 case has been missed at every time unit. I am happy to consider other heuristics.
* Concerning the maximum of 100 time units for backimputation, this should be an upper bound that should work for most cases. There are two reasons. Firstly, if $M$ is the maximum possible generation interval (GI), the $R_{t}$ estimates are not affected by the incidence at time $t-M$. I would assume that for such delays, the GI pmf $\approx 0$  for most cases of interest (unless we specify generation interval distributions at the hour level). Secondly, when back-imputation is performed, the number of imputed cases will decrease exponentially further into the past. So even a growth rate of (1-p/100) implies that the number of cases at time -100 is $(1 - p/100)^{100} \approx e^{-p}$ the number of cases at time $0$. For p=4, this means that the imputed cases at time -100 are about $2\%$ of the cases at time 0.

annecori commented on R/estimate_R.R:
```diff
@@ -190,30 +198,30 @@
 #' ## when not specifying t_start and t_end in config, they are set to estimate
-#' ## the reproduction number on sliding weekly windows                          
-#' res <- estimate_R(incid = Flu2009$incidence, 
+#' ## the reproduction number on sliding weekly windows                      
```
at the end of the examples could you please add one example of how to use back imputation - if possible on one of the existing datasets in the package? @abriz97 . Not totally necessary but will definitely improve visibility of the method I think

abriz97 replied:
answered in commit f00b67e
 [add] example of back_imputation in roxy doc

annecori commented on R/estimate_R.R:
```diff
@@ -28,59 +28,65 @@
 #' @param method One of "non_parametric_si", "parametric_si", "uncertain_si",
 #'   "si_from_data" or "si_from_sample" (see details).
 #'
+#' @param backimputation_window Length of the window used to impute incidence
```
@sangeetabhatia03  should the new argument be added as last argument to ensure full back-compatibility in case people have not specified parameters i.e. are using this as estimate_R(a, b, c) rather than estimate_R(a = a, etc)?

sangeetabhatia03 replied:
agree that would be ideal.

annecori replied:
@abriz97 can you please make that change? thanks

annecori commented:
Thanks @abriz97  for adding your great work to this and @sangeetabhatia03  for reviewing. 
I have tried to sort some issues (mostly unrelated to this PR), and have added a few questions I'd like answers to before merging. 
I also have two more general points:
- @abriz97  I suggest you add a note i main EpiEstim Vignette to point to your specific vignette - to increase visibility of your work
- I wanted to check whether this works with disaggregated data. Idea was to try it out with constant incidence which is reported only weekly – and check how this works atm. I have not had time. Two options: you try it out @abriz97  or if you don't have time you produce an error if someone tries to do this i.e. use your methods with dt > 1, and put in an issue for me or Rebecca to test this in the future. 

Thanks!

abriz97 commented:
Hi @annecori. 

Thank you for your helpful comments. I believe I addressed all of them, and summarise my main changes here:
* I improved "visibility" of my method, by including examples in the documentation and in the main vignette.
* I added an error to stop an user to use the backimputation jointly with Rebecca's disaggregation. I opened an issue (#168) where I suggest an idea I quickly thought about.
* I mentioned the `safe_shift` and maximum imputation delay in the documentation of `backimpute_I`. I am happy to discuss this further.

I also noted previous errors with the implementation, where passing in a data.frame would result in an error. I fixed this and added relevant tests. 

Let me know if there's anything else to cover,

Best,
Andrea

abriz97 commented:
## Further issues

I think we should specify a version of `patchwork` in the `DESCRIPTION` (?) file.
The reason is detailed below.
When I try to run some examples locally, I get the following Error:

```{R}
#> plot(res, "all")
#> Error in Ops.data.frame(guide_loc, panel_loc) : 
#>   ‘==’ only defined for equally-sized data frames
```

This exclusively occurs if I use the "all" argument.
I tracked the error down to the `R/plot.R` file, where we return the following:

```{R}
plot <- p1 + p2 + p3 + plot_layout(ncol = 1)
```

[The issue](https://github.com/thomasp85/patchwork/issues/342) comes from the 
fact I am using an version 1.1.2 of `patchwork`, which is not the last one.
However, I see nowhere in the repository a requirement for this version,
and I believe this should be added.
